python/lsst/daf/butler/_butler.py

```diff
@@ -1862,44 +1862,83 @@ def ingest(
         """
         if not self.isWriteable():
             raise TypeError("Butler is read-only.")
+        if run is not None:
+            warnings.warn(
+                "All DatasetRefs to be ingested must have resolved dataset IDs. "
+                "The run from the ref will be used for ingest and the provided run will be ignored."
+            )
+        log.verbose("Ingesting %d file dataset%s.", len(datasets), "" if len(datasets) == 1 else "s")
+
         progress = Progress("lsst.daf.butler.Butler.ingest", level=logging.DEBUG)
-        # Reorganize the inputs so they're grouped by DatasetType and then
-        # data ID.  We also include a list of DatasetRefs for each FileDataset
-        # to hold the resolved DatasetRefs returned by the Registry, before
-        # it's safe to swap them into FileDataset.refs.
-        # Some type annotation aliases to make that clearer:
-        GroupForType = Dict[DataCoordinate, Tuple[FileDataset, List[DatasetRef]]]
-        GroupedData = MutableMapping[DatasetType, GroupForType]
-        # The actual data structure:
-        groupedData: GroupedData = defaultdict(dict)
+
+        # We need to reorganize all the inputs so that they are grouped
+        # by dataset type and run. Multiple refs in a single FileDataset
+        # are required to share the run and dataset type.
+        GroupedData = MutableMapping[tuple[DatasetType, str], list[FileDataset]]
+        groupedData: GroupedData = defaultdict(list)
+
+        # Track DataIDs that are being ingested so we can spot issues early
+        # with duplication. Retain previous FileDataset so we can report it.
+        groupedDataIds: MutableMapping[
+            tuple[DatasetType, str], dict[DataCoordinate, FileDataset]
+        ] = defaultdict(dict)
+
         # And the nested loop that populates it:
         for dataset in progress.wrap(datasets, desc="Grouping by dataset type"):
-            # This list intentionally shared across the inner loop, since it's
-            # associated with `dataset`.
-            resolvedRefs: List[DatasetRef] = []
-
             # Somewhere to store pre-existing refs if we have an
             # execution butler.
             existingRefs: List[DatasetRef] = []
 
+            dataset_run: str | None = None
             for ref in dataset.refs:
-                if ref.dataId in groupedData[ref.datasetType]:
+                if ref.id is None:
```

Comment:
Open question: If this is an unresolved ref, should I resolve it using `self.run` or should it be an error?

Reply:
Since this question is only relevant during the deprecation period, I think for now we resolve it using the now-deprecated `run` argument, and fall back to `self.run` when that isn't provided.

Reply:
Okay. I now do resolve the refs -- I was forced into this a bit because the resource gathering graph doesn't use resolved refs and so pipeline execution fails because the `_log` files can't be ingested.